src/windowing/window.hpp

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <GLFW/glfw3.h>
+#include <types.hpp>
+#include <ui/movable_rect.hpp>
+
+namespace nebula {
+  struct Window {
+  public:
+    /* Variables */
+    GLFWwindow* glfw_window; // GLFW window instance
+    i32 width; // window width
+    i32 height; // window height
+    void (*custom_render)(); // function that renders all custom objects
+
+    /**
+     * Constructor initializes glfw and creates new fullscreen window with
+     * callbacks.
+     * @param custom_render_func - function that renders all custom objects.
+     * ( Other then MR buttons and created by them rectangles. )
+     */
+    explicit Window(void (*custom_render_func)());
+
+    /**
+     * Destroys window and terminates glfw.
+     */
+    ~Window();
+
+    /* Methods */
+
+    /**
+     * Make this window current context and initializes glad.
+     * @return 0 on success else 1.
+     */
+    [[nodiscard]] u8 make_current_context() const;
+
+    /**
+     * Function to render all objects.
+     */
+    void render() const;
+
+  private:
+    /* Private Variables */
+
+    Vec2 last_mouse_position; // last mouse position
+    List<Movable_Rect> movable_rectangles; // All created rectangles/gates
+    Movable_Rect* currently_moved_rectangle; // Currently moved rect/gate
+
+    /* Private Methods */
+
+    /**
+     * Keyboard input handler.
+     */
+    void keyboard_handler(int key, int scancode, int action, int mods);
+
+    /**
+     * Mouse input handler.
+     */
+    void mouse_handler(int button, int action, int mods);
+
+    /**
+     * Cursor movement handler.
+     */
+    void cursor_handler(double xpos, double ypos);
+
+    /**
+     * Adds new rectangle. Function used by MR buttons.
+     * @param render_function - render function inherited from MR button.
+     * @param rectangle_dimensions - dimensions inherited from MR button.
+     */
+    void add_rectangle(void (*render_function)(Node_Rect),
+                       Vec2 rectangle_dimensions);
+
+    /** Static methods
+     * These are template callback functions redirecting to non-static handlers.
```

Comment:
purpose of these? why can't those be hidden in the source file as static linkage non-member functions?